catboost/libs/algo/mvs.cpp

```diff
@@ -18,22 +18,110 @@ inline static double GetSingleProbability(double derivativeAbsoluteValue, double
     return (derivativeAbsoluteValue > threshold) ? 1.0 : (derivativeAbsoluteValue / threshold);
 }
 
+inline static double CalculateLastIterMeanLeafValue(const TVector<TVector<TVector<double>>>& leafValues) {
+    const ui32 lastIter = leafValues.ysize();
+    const TVector<double>& lastIterValues = leafValues[lastIter - 1][0]; // one dimensional approx
+    double sumOfAbsoluteValues = 0.0;
+    for (double value : lastIterValues) {
+        sumOfAbsoluteValues += Abs(value);
+    }
+    return sumOfAbsoluteValues / lastIterValues.ysize();
+}
+
+static double CalculateMeanGradValue(const double* derivatives, ui32 cnt, NPar::TLocalExecutor* localExecutor) {
+        NPar::TLocalExecutor::TExecRangeParams blockParams(0, cnt);
+        blockParams.SetBlockCount(CB_THREAD_LIMIT);
+        TVector<double> gradSumInBlock(blockParams.GetBlockCount(), 0.0);
+        localExecutor->ExecRange(
+            [&](ui32 blockId) {
+                const ui32 blockOffset = blockId * blockParams.GetBlockSize();
+                const ui32 blockSize = Min(
+                    static_cast<ui32>(blockParams.GetBlockSize()),
+                    cnt - blockOffset
+                );
+                const ui32 blockFinish = blockOffset + blockSize;
+                const double* derivatives_begin = derivatives + blockOffset;
+                const double* derivatives_end = derivatives + blockFinish;
+                for (auto it = derivatives_begin; it != derivatives_end; ++it) {
+                    gradSumInBlock[blockId] += Abs(*it);
+                }
+            },
+            0,
+            blockParams.GetBlockCount(),
+            NPar::TLocalExecutor::WAIT_COMPLETE
+        );
+
+        const double sum_gradients = Accumulate(gradSumInBlock.begin(), gradSumInBlock.end(), 0.0);
+        return sum_gradients / cnt;
+}
+
+double TMvsSampler::GetLambda(
+    const double* derivatives,
```

Comment:
style guilde again - output variables should be last ones among other parameters.
Also if something is changed inside the function (in this case derivatives), it should be obvious from the interface. In this case it's not clear.

Reply:
I guess, you did not notice "const". Or is the pointer to constant an output variable?

Reply:
You're right, I didn't notice const, closing this one.